app/models/game.rb

```diff
@@ -29,29 +29,210 @@ def populate_board!
-
-    # Store enemy_king coordinates for shorter reference
-    ekx = enemy_king.x_coordinate
-    eky = enemy_king.y_coordinate
-
-    player_pieces.each do |player_piece|
-      # Check if it would be valid for the current player_piece to move to the enemy king's position
-      if player_piece.valid_move? ekx, eky
-        # Check if the current player_piece wouldn't obstructed while attempting to move to the enemy king's position
-        unless player_piece.is_obstructed? ekx, eky
-          # If both criteria are met, at least once player_piece is successfully 'checking' the enemy king
+    set_remaining_piece_data color
+    
+    @remaining_player_pieces.each do |player_piece|
+    # Check if it would be valid for the current player_piece to move to the opponent king's position
+      if player_piece.valid_move? @okx, @oky
+      # Check if the current player_piece wouldn't obstructed while attempting to move to the opponent king's position
+        unless player_piece.is_obstructed? @okx, @oky
+        # Store reference to player piece causing check
+          @player_piece_causing_check = player_piece
+        # Store player_piece coordinates for shorter reference elsewhere
+          @x_of_threat = player_piece.x_coordinate
+          @y_of_threat = player_piece.y_coordinate
+        # If both criteria are met, at least one player_piece is 'checking' the opponent king
           return true
         end
       end
     end
-    # Return false if no player_piece is 'checking' the enemy king
+
+  # Return false: no player_piece is 'checking' the opponent king
+    false
+  end
+  
+  def ends_by_checkmate?(color)
+  # Return false immediately unless the game is in check
+    return false unless in_check? color
+  # Return false if opponent can capture the piece(s) causing check
+    return false if opponent_can_capture_piece_causing_check? color
+  # Return false if opponent can block the piece(s) causing check
+    return false if opponent_can_block_piece_causing_check? color
+  # Return false if opponent king can escape check
+    return false if checked_king_can_escape?
+
+    true
+  end
+  
+  def opponent_can_capture_piece_causing_check?(color)
+    @remaining_opponent_pieces.each do |opponent_piece|
+    # Skip this iteration if the opponent_piece is a pawn that couldn't capture the piece causing check
+      if opponent_piece.type.eql? 'Pawn'
+        next if ( opponent_piece.x_coordinate - @x_of_threat ).abs != 1 || ( opponent_piece.y_coordinate - @y_of_threat ).abs != 1
+      end
+      
+    # Check if it would be valid for the opponent_piece to move to the position of the piece causing check
+      if opponent_piece.valid_move? @x_of_threat, @y_of_threat
+      # Check if the opponent_piece wouldn't obstructed while attempting to move to the position of the piece causing check
+        unless opponent_piece.is_obstructed? @x_of_threat, @y_of_threat
```

Comment:
You could use 'next if' in place of 'unless' and save another line by removing the 'end' on line 80.